### PR TITLE
Drivers: Flash: stm32: G0 dual bank handling, G0 error check and CFGBSY check fixes

### LINKED ISSUE #37389: nucleo_g0b1re: Swapping image in mcuboot results in hard fault and softbricks the device
**Describe the bug**
On STM32G0B1RE devices, `mcuboot` image upgrade (swapping) would result in hard fault and the device will not be able to boot again. This initially occurs on my g0b1 custom board, which can be debugged using a Segger Jlink. I also verify this on the `nucleo_g0b1` tree board, which ended in the same condition albeit can't verify that it is also a hard fault as there seems to be no easy way to debug it.

**To Reproduce**
1. Build and flash the mcuboot bootloader on the `nucleo_g0b1re` board
2. Build and flash the `mgmt/smp_svr` sample on the `nucleo_g0b1re` board
3. Use `mcumgr` to upload a `hello_world` image and perform `image test`
4. Reset the device so that it performs upgrade on the next boot

You should see something like this:

```
*** Booting Zephyr OS build zephyr-v2.6.0-1636-g549ecbf6ca46  ***
[00:00:00.005,000] <inf> mcuboot: Starting bootloader
[00:00:00.007,000] <inf> mcuboot: Primary image: magic=good, swap_type=0x1, copy_done=0x3, image_ok=0x1
[00:00:00.007,000] <inf> mcuboot: Scratch: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
[00:00:00.007,000] <inf> mcuboot: Boot source: primary slot
[00:00:00.021,000] <inf> mcuboot: Swap type: test
*** Booting Zephyr OS build zephyr-v2.6.0-1636-g549ecbf6ca46  ***
[00:00:00.005,000] <inf> mcuboot: Starting bootloader
[00:00:00.007,000] <inf> mcuboot: Primary image: magic=good, swap_type=0x1, copy_done=0x3, image_ok=0x1
[00:00:00.007,000] <inf> mcuboot: Scratch: magic=good, swap_type=0x2, copy_done=0x3, image_ok=0x3
[00:00:00.007,000] <inf> mcuboot: Boot source: scratch
```

And it just stay here forever and can't boot into the bootloader (bootloader likely corrupted) anymore even after reset. If the bootloader is reflashed (but the application isn't), then it will repeat the whole thing again. To boot into the bootloader again, a full chip erase is required.

After some debugging the hard fault seems to occurs at when the bootloader is erasing the flash of the device [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/flash/flash_stm32g0x.c#L104).

**Expected behavior**
The mcuboot should be able to finish the swapping of the image.

**Impact**
Unable to perform DFU.

**Environment (please complete the following information):**
 - OS: Windows 10
 - Toolchain: GNUARMEMB
 - v2.6.0-1636-g549ecbf6ca46

**Additional context**
To compile the bootloader for the nucleo board, I declared a scratch partition in the dts like so:
```
&flash0 {
	partitions {
		compatible = "fixed-partitions";
		#address-cells = <1>;
		#size-cells = <1>;

		boot_partition: partition@0 {
			label = "mcuboot";
			reg = <0x00000000 0x00010000>;
			read-only;
		};
		slot0_partition: partition@10000 {
			label = "image-0";
			reg = <0x00010000 0x00032800>;
		};
		slot1_partition: partition@42800 {
			label = "image-1";
			reg = <0x00042800 0x00032800>;
		};
		storage_partition: partition@75000 {
			label = "storage";
			reg = <0x00075000 0x0003000>;
		};
		scratch_partition: partition@78000 {
			label = "image-scratch";
			reg = <0x00078000 0x0008000>;
		};
	};
};
```

The `smp_svr` image is signed and confirmed.

### PR DESCRIPTION
**Pull request by str4t0m**
This PRs fixes several issues in the stm32g0 flash driver and introduces the checking of the  CFGBSY bit in the wait_flash_idle routine for all series supporting it.
Furthermore the integration_platforms for the drivers/flash test case are modified, such that several stm32 boards are build in CI.

Regarding dual bank flash bank layout on STM32G0:
This has only been tested on the g0b1re 512KiB flash target.
I am not sure about the influence of the DUAL_BANK option on 256KiB targets.
Does anybody of the ST guys know which SoC has FLASH_SALES_TYPE_0?
This seems to be the only type for which the DUAL_BANK option seems to have an influence on the number of banks (FLASH_BANK_NB)

Closes #37389

### CODE REVIEW

**Review by ABOSTM — CHANGES_REQUESTED** [review]
@str4t0m , 
Thanks a lot for this PR, that is a great job.
I don't have the information about FLASH_SALES_TYPE (undocumented register), but what I understand is that we don't really need it.Chips with flash 256KB and more, have the support of Dualbank. So we can guess the CHIPs: G0BxxE G0BxxC , G0C1xE, G0C1xC have 256KB or more support dualbank,  ; CHIPs: G0B1xB is 128KB is single bank only, 
Thus what you have done looks good to me (CONFIG_FLASH_SIZE > (128))
I even wonder why HAL uses an undocumented register, whereas documented FLASH_SIZE register could have been used instead.

That said, when flash is 256K or more, dualbank is supported.
But it doesn't mean it is enabled:
this is the interest of option byte  DUAL_BANK : 
if set (default state out of the box) flash is configured in dualbank
if cleared, then flash is single bank (continuous).
In stm32g4 flash driver, we decided for now to not support it. but at least it is checked:
```
#if defined(FLASH_OPTR_DBANK)
	if (READ_BIT(FLASH->OPTR, FLASH_OPTR_DBANK) == 0U) {
		/* Single bank not supported when dualbank is possible */
		LOG_ERR("Single bank configuration not supported");
		return -ENOTSUP;
	}
#endif
```
So for G0 either you could do the same (check the option byte, but don't support single bank configuration)
Or fully support dualbank/single bank configuration, for the same MCU .
Up to you. You already made a great job, not supporting it would be fine.

**Comment by str4t0m** [comment]
Thanks for the review.
One reason why they could have chosen to use the flash sales type could be that the behavior has changed between revisions, but the naming of this register does not  really affirm something like that.

Regarding the dual bank option byte:
I have not been able to observe any differences on G0B1RE, whether this was set or not.
Basic flash operations and mcuboot updates with two image slots work on both configurations.
The FLASH_BANK_NB macro always returns 2 for G0B1RE, so most likely it only returns 1 for 256kiB SoCs when additionally OB_DUAL_BANK_VALUE is 0.
Would it be fine if -ENOTSUP is only returned in case OB_DUAL_BANK_VALUE is 0 and additionally CONFIG_FLASH_SIZE==256?
I have done it so in the latest update.

File: drivers/flash/flash_stm32.h [hunk 1]
```diff
@@ -33,9 +33,43 @@ struct flash_stm32_priv {
 	/* as flash node property 'write-block-size' */
 #endif
 
+/* Differentiate between arm trust-zone non-secure/secure, and others. */
+#if defined(FLASH_NSSR_NSBSY)		/* For mcu w. TZ in non-secure mode */
+#define FLASH_SECURITY_NS
```
[the remark this answers is not in the document]
Comment (str4t0m) [reply drivers/flash/flash_stm32.h:38]: Agree, as there won't be added much more registers using extra ifdefs is superfluous.

**Comment by ABOSTM** [comment]
I can imagine that the effect of dual bank option byte, could be seen on mass erase operation:
* in single bank, the whole flash would be erased
* in dualbank, only half of the flash is erased (FLASH_CR_MER1)

Then I agree with you that G0B1RE always return FLASH_BANK_NB  =2 whatever the option byte.
This sounds to me like a bug (probably stm32 HAL bug), thus I submitted an internal issue. 
So I don't think it is good to return -ENOTSUP only for CONFIG_FLASH_SIZE==256
But let's wait for the expert feedback. I set DNM label while waiting.

**Comment by ycsin** [comment]
I have these changes in my `hawkbit` testing and it has been working great so far, I can confirm that it fixed #37389

**Comment by erwango** [comment]
> But let's wait for the expert feedback. I set DNM label while waiting.

Internal ref 111811. No update.

**Comment by ycsin** [comment]
Any update?

**Comment by ycsin** [comment]
ping @str4t0m @ABOSTM @erwango 

maybe we should just get this through if there's no update from ST's expert? Personally I'd rather have some potential bug than the flash driver doesn't work at all for G0B1RE.

**Review by erwango — APPROVED** [review]
I agree with @ycsin.
This PR brings more benefit than the risk.
Let's get it merged, if required, let's enter an issue to remind of the pending point

**Review by ycsin — APPROVED** [review]
Thanks @str4t0m for this amazing work, haven't had any issue so far on my boards

**Review by ABOSTM — APPROVED** [review]
OK for a merge even if we don't have the final answer.

**Comment by ycsin** [comment]
@erwango is this PR eligible to be back ported to v2.7?

**Pull request merged**